cmd/juju/application/config.go

```diff
@@ -117,34 +118,104 @@ func (c *configCommand) Init(args []string) error {
+	// strings and verify them. If there is one or more valid keys to reset and
+	// no other errors initalizing the command, c.resetDefaults will be called
+	// in c.Run.
+	if err := c.parseResetKeys(); err != nil {
+		return errors.Trace(err)
+	}
+
 	c.applicationName = args[0]
-	if c.reset {
-		c.action = c.resetConfig
-		return c.parseReset(args[1:])
+	args = args[1:]
+
+	switch len(args) {
+	case 0:
+		return c.handleZeroArgs()
+	case 1:
+		return c.handleOneArg(args)
+	default:
+		return c.handleArgs(args)
 	}
+}
+
+// handleZeroArgs handles the case where there are no positional args.
+func (c *configCommand) handleZeroArgs() error {
+	// If there's a path we're setting args from a file
 	if c.configFile.Path != "" {
-		return c.parseSet(args[1:], true)
+		return c.parseSet([]string{})
+	}
+	if len(c.reset) == 0 {
+		// If there's nothing to reset we're getting all the settings.
+		c.action = c.getConfig
+	}
+	// Otherwise just reset.
+	return nil
+}
+
+// handleOneArg handles the case where there is one positional arg.
+func (c *configCommand) handleOneArg(args []string) error {
+	// If there's an '=', this must be setting a value
+	if strings.Contains(args[0], "=") {
+		return c.parseSet(args)
+	}
+	// If there's no reset,	we want to get a single value
+	if len(c.reset) == 0 {
+		c.action = c.getConfig
+		c.keys = args
+		return nil
+	}
+	// Otherwise we have reset and a get arg, which is invalid.
+	return errors.New("cannot reset and retrieve values simultaneously")
+}
+
+// handleArgs handles the case where there's more than one positional arg.
+func (c *configCommand) handleArgs(args []string) error {
+	// This must be setting values but let's make sure.
+	var pairs, numArgs int
+	numArgs = len(args)
+	for _, a := range args {
+		if strings.Contains(a, "=") {
+			pairs++
+		}
+	}
+	if pairs == numArgs {
+		return c.parseSet(args)
 	}
-	if len(args[1:]) > 0 && strings.Contains(args[1], "=") {
-		return c.parseSet(args[1:], false)
+	if pairs == 0 {
+		return errors.New("can only retrieve a single value, or all values")
 	}
-	return c.parseGet(args[1:])
+	return errors.New("cannot set and retrieve values simultaneously")
 }
 
-// parseReset parses command line args when the --reset flag is supplied.
-func (c *configCommand) parseReset(args []string) error {
-	if len(args) == 0 {
-		return errors.New("no configuration options specified")
+// parseResetKeys splits the keys provided to --reset after trimming any
+// leading or trailing comma. It then verifies that we haven't incorrectly
+// received any key=value pairs and finally sets the value(s) on c.resetKeys.
+func (c *configCommand) parseResetKeys() error {
+	if len(c.reset) == 0 {
+		return nil
+	}
+	var resetKeys []string
+	for _, value := range c.reset {
+		keys := strings.Split(strings.Trim(value, ","), ",")
+		resetKeys = append(resetKeys, keys...)
+	}
+	for _, k := range resetKeys {
+		if strings.Contains(k, "=") {
```

Comment:
This is a good point. We could fail faster and there wouldn't be any loss of fidelity in the error message. However, a PR with this has already landed. I'd say if you think it is important enough we should open a bug and update all three items that now use `--reset a,b,c`.